### PR TITLE
[BUGFIX] Fix Snowflake `'connection_string'` `TestConnectionError`

### LINKED ISSUE #8731: Snowflake datacontext with connection parameter throws an error.
**Describe the bug**
snowflake datacontext with connection parameter has defect?
connection_string should be optional if we pass connection parameters like user, password, schema, role, db.

**To Reproduce**

1. Get the gx context
2. Add new datasource connection using context.sources.add_snowflake()
3. Instead of using connection_string, try to add connection parameters like, user, password, account, database, schema, warehouse and role

**Expected behavior**
It should add new snowflake datasource

**Environment (please complete the following information):**
 - Operating System: MacOS
 - Great Expectations Version: great_expectations==0.17.16
 - Data Source: Snowflake
 - Cloud environment: AWS

**Additional context**
Error message
```
Traceback (most recent call last):
  File "/Users/stavan.mehta/projects/spaceship/spaceship/lib/python3.11/site-packages/great_expectations/datasource/fluent/snowflake_datasource.py", line 133, in get_engine
    connection_string = model_dict.pop("connection_string")
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
KeyError: 'connection_string'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/stavan.mehta/projects/spaceship/spaceship/lib/python3.11/site-packages/great_expectations/datasource/fluent/sql_datasource.py", line 1069, in test_connection
    engine: sqlalchemy.Engine = self.get_engine()
                                ^^^^^^^^^^^^^^^^^
  File "/Users/stavan.mehta/projects/spaceship/spaceship/lib/python3.11/site-packages/great_expectations/datasource/fluent/snowflake_datasource.py", line 143, in get_engine
    raise SQLDatasourceError(
great_expectations.datasource.fluent.sql_datasource.SQLDatasourceError: Unable to create a SQLAlchemy engine from connection_string: None due to the following exception: 'connection_string'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/stavan.mehta/projects/spaceship/gx_cloud_setup.py", line 24, in <module>
    datasource = context.sources.add_snowflake(
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/stavan.mehta/projects/spaceship/spaceship/lib/python3.11/site-packages/great_expectations/datasource/fluent/sources.py", line 472, in add_datasource
    datasource.test_connection()
  File "/Users/stavan.mehta/projects/spaceship/spaceship/lib/python3.11/site-packages/great_expectations/datasource/fluent/sql_datasource.py", line 1072, in test_connection
    raise TestConnectionError(
great_expectations.datasource.fluent.interfaces.TestConnectionError: Attempt to connect to datasource failed with the following error message: Unable to create a SQLAlchemy engine from connection_string: None due to the following exception: 'connection_string'
```

### PR DESCRIPTION
**Pull request by Kilo59**
Fix an issue that prevents using a Snowflake datasource if it was created without using a `connection_string`.

Adds tests to prevent regression.

### CODE REVIEW

File: tests/datasource/fluent/test_snowflake_datasource.py [hunk 1]
```diff
@@ -11,6 +12,21 @@
...
+        {"connection_string": "snowflake://my_user:password@my_account"},
+        {"user": "my_user", "password": "password", "account": "my_account"},
+    ],
+)
+def test_valid_config(config_kwargs: dict):
+    my_sf_ds_1 = SnowflakeDatasource(name="my_sf_ds_1", **config_kwargs)
+    assert my_sf_ds_1
+
+    sql_engine = my_sf_ds_1.get_engine()
+    assert isinstance(sql_engine, sa.engine.Engine)
```
Comment (Kilo59) [inline tests/datasource/fluent/test_snowflake_datasource.py:28]: This test fails without the applied fix.

File: tests/datasource/fluent/test_snowflake_datasource.py [hunk 2]
```diff
@@ -127,3 +144,7 @@ def test_get_execution_engine_succeeds():
     )
     # testing that this doesn't raise an exception
     datasource.get_execution_engine()
+
+
+if __name__ == "__main__":
+    pytest.main([__file__, "-vv"])
```
Comment (anthonyburdi) [inline tests/datasource/fluent/test_snowflake_datasource.py:150]: Is this just for local testing or did you mean to include?

**Review by anthonyburdi — APPROVED** [review]
LGTM just one comment for clarification.

File: tests/datasource/fluent/test_snowflake_datasource.py (hunk 2 above)
Comment (Kilo59) [reply tests/datasource/fluent/test_snowflake_datasource.py:150]: It's to make running the tests for this single module easier (particularly from within an IDE).

**Pull request merged**